src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs

```diff
@@ -0,0 +1,872 @@
+            }
+
+            ProduceStart(appCompleted: false);
+        }
+
+        private void ThrowResponseAbortedException()
+        {
+            throw new ObjectDisposedException("Unhandled application exception", _applicationException);
+        }
+
+        private void ProduceStart(bool appCompleted)
+        {
+            if (HasResponseStarted)
+            {
+                return;
+            }
+
+            HasResponseStarted = true;
+
+            SendResponseHeaders(appCompleted);
+
+            StartWritingResponseBody();
+        }
+
+        protected Task ProduceEnd()
+        {
+            if (_applicationException != null)
+            {
+                if (HasResponseStarted)
+                {
+                    // We can no longer change the response, so we simply close the connection.
+                    return Task.CompletedTask;
+                }
+
+                // If the request was rejected, the error state has already been set by SetBadRequestState and
+                // that should take precedence.
+                else
+                {
+                    // 500 Internal Server Error
+                    SetErrorResponseHeaders(statusCode: StatusCodes.Status500InternalServerError);
+                }
+            }
+
+            if (!HasResponseStarted)
+            {
+                return ProduceEndAwaited();
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private void SetErrorResponseHeaders(int statusCode)
+        {
+            StatusCode = statusCode;
+        }
+
+        private async Task ProduceEndAwaited()
+        {
+            ProduceStart(appCompleted: true);
+
+            // Force flush
+            await Output.FlushAsync();
+        }
+
+        public unsafe void SendResponseHeaders(bool appCompleted)
+        {
+            // Verifies we have sent the statuscode before writing a header
+            var reasonPhrase = ReasonPhrases.GetReasonPhrase(StatusCode);
+            var reasonPhraseBytes = Encoding.UTF8.GetBytes(reasonPhrase);
+
+            fixed (byte* pReasonPhrase = reasonPhraseBytes)
+            {
+                // This copies data into the underlying buffer
+                NativeMethods.http_set_response_status_code(_pHttpContext, (ushort)StatusCode, pReasonPhrase);
+            }
+
+            HttpApiTypes.HTTP_RESPONSE_V2* pHttpResponse = NativeMethods.http_get_raw_response(_pHttpContext);
+
+            // We should be sending the headers here as there are many responses that don't have status codes.
+            _pinnedHeaders = SerializeHeaders(pHttpResponse);
+        }
+
+        // From HttpSys, except does not write to response
+        private unsafe List<GCHandle> SerializeHeaders(HttpApiTypes.HTTP_RESPONSE_V2* pHttpResponse)
+        {
+            // TODO make the header collection readonly to stop modifications.
+            HttpApiTypes.HTTP_UNKNOWN_HEADER[] unknownHeaders = null;
```

Comment:
From HttpSys, except we don't send the header response here. Could maybe consider some code sharing in the future, however we need to be careful about buffer management.